client/state/routing/middleware.js

```diff
@@ -26,8 +31,9 @@ export const restoreLastLocation = () => {
 			( lastPath ) => {
 				if ( ! hasInitialized &&
 						lastPath && lastPath !== '/' &&
-						action.path === '/' &&
-						! isOutsideCalypso( lastPath ) ) {
+						action.path === '/' && Object.keys( action.query ).length === 0 &&
```

Comment:
I added the check to the same place where we are checking for `action.path` existence. Seems that all `ROUTE_SET` actions should have both of those properties present.